### DISCUSSION TITLE
Running Credential fetcher in Ubuntu environment

### DISCUSSION THREAD

**Discussion by Sachyyy** [aws/credentials-fetcher#81] [Q&A]
Hello,
I am trying to run Credential fetcher on ubuntu environment.
This is a AD joined linux environment.

I tried following the credentials-fetcher configuration guide but failed getting Kerberos Ticket for the gMSA account on the linux machine.

>  sudo grpc_cli call unix:/var/credentials-fetcher/socket/credentials_fetcher.sock AddKerberosLease "credspec_contents: '{"CmsPlugins":["ActiveDirectory"],"DomainJoinConfig":{"Sid":"S-1-5-21-3036598194-874362413-4262823843","MachineAccountName":"MGSA_DevTS2CSvc","Guid":"6e23677c-00a8-4a89-a589-0ef2847291bf","DnsTreeName":"dev.local","DnsName":"dev.local","NetBiosName":"DEV"},"ActiveDirectoryConfig":{"GroupManagedServiceAccounts":[{"Name":"MGSA_DevTS2CSvc","Scope":"dev.local"},{"Name":"MGSA_DevTS2CSvc","Scope":"DEV"}]}}'"
Method name not found
Failed to find method AddKerberosLease in proto files.
Method name not found
Method name not found
Failed to parse request.
Reflection request not implemented; is the ServerReflection service enabled?

I am working with my colleague who also post another discussion on correctly installing Credential fetcher on Ubuntu.

**Comment by sb-ruisms** [comment]
Thanks for posting this. Will look into it and keep this thread posted.

**Comment by saikiranakula-amzn** [comment]
Hi Sachyyy@, Thank you for reaching out to us. When using grpc cli. The formatting need to be very similar to the example provided below.
`
sudo grpc_cli call unix:/usr/share/credentials-fetcher/socket/credentials_fetcher.sock AddKerberosLease "credspec_contents: '{\"CmsPlugins\":[\"ActiveDirectory\"],\"DomainJoinConfig\":{\"Sid\":\"S-1-5-21-4217655605-3681839426-3493040985\",\"MachineAccountName\":\"WebApp01\",\"Guid\":\"af602f85-d754-4eea-9fa8-fd76810485f1\",\"DnsTreeName\":\"contoso.com\",\"DnsName\":\"contoso.com\",\"NetBiosName\":\"contoso\"},\"ActiveDirectoryConfig\":{\"GroupManagedServiceAccounts\":[{\"Name\":\"WebApp01\",\"Scope\":\"contoso.com\"},{\"Name\":\"WebApp01\",\"Scope\":\"contoso\"}]}}'" 
`
Also, as a sanity check, ensure the credentials-fetcher is running on the host "sudo systemctl status credentials-fetcher", and check the credentials-fetcher.sock is available at "/var/credentials-fetcher/socket"
please try this and keep us posted if this resolves the issue

**Comment by smhmhmd** [comment]
@Sachyyy 
For initial testing, you can also use the test-client.
The test code uses account names `webapp01` and `webapp02` and`contoso.com` as domain.
```
$ ./api/tests/gmsa_test_client

**Comment by Sachyyy** [comment]
I did the escaping with the example you provided, still the same. 

> sudo grpc_cli call unix:/var/credentials-fetcher/socket/credentials_fetcher.sock AddKerberosLease "credspec_contents: '{\"CmsPlugins\":[\"ActiveDirectory\"],\"DomainJoinConfig\":{\"Sid\":\"S-1-5-21-3036598194-874362413-4262823843\",\"MachineAccountName\":\"MGSA_DevTS2CSvc\",\"Guid\":\"6e23677c-00a8-4a89-a589-0ef2847291bf\",\"DnsTreeName\":\"dev.local\",\"DnsName\":\"dev.local\",\"NetBiosName\":\"DEV\"},\"ActiveDirectoryConfig\":{\"GroupManagedServiceAccounts\":[{\"Name\":\"MGSA_DevTS2CSvc\",\"Scope\":\"dev.local\"},{\"Name\":\"MGSA_DevTS2CSvc\",\"Scope\":\"DEV\"}]}}'"

Also Credetials Fetcher is running
> ● credentials-fetcher.service - credentials-fetcher systemd service unit file.
     Loaded: loaded (/etc/systemd/system/credentials-fetcher.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2023-11-07 09:47:20 UTC; 2 weeks 0 days ago
   Main PID: 713 (credentials-fet)
     Status: "Watchdog notify count = 257241"
      Tasks: 12 (limit: 9388)
     Memory: 16.0M
        CPU: 2min 56.267s
     CGroup: /system.slice/credentials-fetcher.service
             └─713 /usr/sbin/credentials-fetcherd

**Reply by Sachyyy** [reply]
Also from the logs, 
It was complaining grpc_cli unable to get the proto definition. 
And the proto reflection service is not enabled. 
how do we come around this?

**Reply by saikiranakula-amzn** [reply]
I would suggest using gMSA test client to test these. grpc cli sometime it is hard to find if there are any unnoticed characters.
Update this string as per your gMSA account and domain information - https://github.com/aws/credentials-fetcher/blob/17b9df0b8e9e5bb59a5cd6fff3d5fc31fa656272/api/tests/gmsa_test_client.cpp#L443, and after you run make. Navigate to  /build/api/test folder run the binary gmsa_test_client to validate the use-case.

**Reply by Sachyyy** [reply]
Hi @saikiranakula-amzn 

I did the changes locally and make and run the gmsa_test_client. 
following was the result.

> 13: ERROR: cannot get machine krb ticket
Client received output for add kerberos lease: RPC failed
Client received output for delete kerberos lease: RPC failed

**Reply by Sachyyy** [reply]
I did trace the above log lines, Wanted to see whats going on. So I added few more log lines. but unfortunately my log lines were not printed.

My steps. 
Change the code,
in the build folder run 'make'
did a restart of credentials_fetcher
run gmsa_test_client in build/api/test folder.

any other steps I should be doing ?

**Reply by Sachyyy** [reply]
@saikiranakula-amzn @smhmhmd can we have a call to resolve this issue and make the gmsa_test_client running ?

**Reply by saikiranakula-amzn** [reply]
Hi sachayy, we can have a quick call today or tomorrow to get the issues fixed

**Reply by saikiranakula-amzn** [reply]
sachayy, can you also please share your email address

**Reply by smhmhmd** [reply]
@Sachyyy - We only have @cskumar6381 's contact, @saikiranakula-amzn has setup a meeting with him, please check with him.

**Reply by Sachyyy** [reply]
Just saw your message. It's jayawarna@gmail.com. Sure I will coordinate with him on the meeting schedule. Thank you!

**Reply by saikiranakula-amzn** [reply]
will send out the invite for today

**Reply by Sachyyy** [reply]
Hi Guys,
@cskumar6381  did resolved that issue of DNS server and after that the ldapsearch is running. But I believe there is something missing here. 

journal shows like this 
>Dec 01 03:39:30 ubuntu-v credentials-fetcherd[7600]: ldapsearch -H ldap://DC1-V.dev.local -b 'CN=MGSA_DevTS2CSvc,CN=Managed Service Accounts,DC=dev,DC=local' -s sub  "(objectClass=msDs-GroupManagedServiceAccount)"  msDS-ManagedPassword
Dec 01 03:39:30 ubuntu-v ldapsearch[7671]: GSSAPI client step 1
Dec 01 03:39:30 ubuntu-v credentials-fetcherd[7671]: SASL/GSS-SPNEGO authentication started
Dec 01 03:39:30 ubuntu-v ldapsearch[7671]: GSSAPI client step 1
Dec 01 03:39:30 ubuntu-v ldapsearch[7671]: GSSAPI client step 1
Dec 01 03:39:30 ubuntu-v credentials-fetcherd[7671]: SASL username: UBUNTU-V$@DEV.LOCAL
Dec 01 03:39:30 ubuntu-v credentials-fetcherd[7671]: SASL SSF: 256
Dec 01 03:39:30 ubuntu-v credentials-fetcherd[7671]: SASL data security layer installed.
Dec 01 03:39:30 ubuntu-v ldapsearch[7671]: DIGEST-MD5 common mech free
Dec 01 03:39:30 ubuntu-v credentials-fetcherd[7600]: ERROR: Cannot get gMSA krb ticket
Dec 01 03:48:18 ubuntu-v credentials-fetcherd[7600]: CallDataCreateKerberosLease 0x7fae1c00c570 status: 2
Dec 01 03:48:18 ubuntu-v credentials-fetcherd[7600]: CallDataDeleteKerberosLease 0x7fae1c00e400 status: 1
Dec 01 03:48:18 ubuntu-v credentials-fetcherd[7600]: CallDataDeleteKerberosLease 0x7fae1c0162c0 status: 0
Dec 01 03:48:18 ubuntu-v credentials-fetcherd[7600]: CallDataDeleteKerberosLease 0x7fae1c00e400 status: 2

But when I run the test client,
still giving an error. does not say anything. 
>13: ERROR: Cannot get gMSA krb ticket
Client received output for add kerberos lease: RPC failed
Client received output for delete kerberos lease: RPC failed

Actually Appreciate if you could help me what did I miss when make, restart the service and running the test, still not showing the logs I included.

**Reply by Sachyyy** [reply]
Hi Guys,
We found out that the ldapsearch still not retrieving the account because inside our folder hierarchy, it could not find it. 
But if you execute the command like this
>ldapsearch -H ldap://DC1-V.dev.local -b 'CN=MGSA_DevTS2CSvc,OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV,DC=dev,DC=local' -s sub  "(objectClass=msDs-GroupManagedServiceAccount)"  msDS-ManagedPassword

actually, it can retrieve the account. 
so we have 2 questions.
1) how do we pass a folder hierarchy to the gmsa_test_client
2) Most importantly, we want to see if the GRPC_CLI can retrieve it so that we could utilize that in our setup. 

appreciate a your reply on this. Thanks !

**Reply by smhmhmd** [reply]
>We found out that the ldapsearch still not retrieving the account because inside our folder hierarchy, it could not find it.
But if you execute the command like this

>ldapsearch -H ldap://DC1-V.dev.local -b 'CN=MGSA_DevTS2CSvc,OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV,DC=dev,DC=local' -s sub "(objectClass=msDs-GroupManagedServiceAccount)" msDS-ManagedPassword

This PR adds a new env variable.
https://github.com/aws/credentials-fetcher/pull/86/commits/459b3e2e74d93e664cc87f29c4815a0c16f8c766

**Reply by Sachyyy** [reply]
Hi @smhmhmd ,
I had to go back to my very first question. I do change the code, but seems it does not the one that runs.
I did change the code so that folder hierarchy is set. 
But it did not pick up.
I change the code to 
>     /**
     * ldapsearch -H ldap://<fqdn> -b 'CN=webapp01,CN=Managed Service
     *   Accounts,DC=contoso,DC=com' -s sub  "(objectClass=msDs-GroupManagedServiceAccount)"
     *   msDS-ManagedPassword
     */
    std::string gmsa_ou = std::string( ",CN=Managed Service Accounts," );
    if ( getenv(ENV_CF_GMSA_OU) != NULL)
    {
        gmsa_ou = std::string( "," ) + getenv(ENV_CF_GMSA_OU)+ std::string( "," );
        cf_logger.logger(LOG_ERR, "sachith test %s", gmsa_ou);
    }
    std::string cmd = std::string( "ldapsearch -H ldap://" ) ;
    cmd += std::string( " -b 'CN=" ) + gmsa_account_name + gmsa_ou + base_dn + std::string( "'" ) +
    std::string( " -s sub  \"(objectClass=msDs-GroupManagedServiceAccount)\" "
                        " msDS-ManagedPassword" );


even I removed, "fqdn" to see if the journal shows the change in command, but unfortunately it runs the same command.
So I need to verify the code built and run properly.

**Reply by smhmhmd** [reply]
Did you set the env var in the service file ?
Please copy your service file here

**Reply by Sachyyy** [reply]
> Environment="CF_GMSA_OU=OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV"

**Reply by smhmhmd** [reply]
I am assuming you set it like this in your service file
``` export CF_GMSA_OU="OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV"```

**Reply by Sachyyy** [reply]
Hi @smhmhmd 
I add the the env like this (which did not have any errors)
> Environment="CF_GMSA_OU=OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV"

But I changed to what you sent me. with export keyword. then the journal was like this.
> Dec 05 04:44:25 ubuntu-v systemd[1]: /etc/systemd/system/credentials-fetcher.service:13: Unknown key name 'export CF_GMSA_OU' in section 'Service', ignoring.

**Reply by smhmhmd** [reply]
I hope ```CF_GMSA_OU="OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV"``` works for you.

**Reply by smhmhmd** [reply]
```CF_GMSA_OU="OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV"```
looks different from 
>"CF_GMSA_OU=OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV"

**Reply by Sachyyy** [reply]
Does not work. Only one worked was
> Environment="CF_GMSA_OU=OU=Dev_MGSAs,OU=Managed Group Service Accounts,OU=System Accounts,OU=DEV"
not sure if I need to escape any characters here.

**Reply by smhmhmd** [reply]
>Does not work. Only one worked was

Do you still have the issue ?

**Reply by smhmhmd** [reply]
@Sachyyy If you still have the issue, lets have a call tomorrow, you have our email

**Reply by Sachyyy** [reply]
Yes. Let's have a call tomorrow.
Thank you !

**Reply by smhmhmd** [reply]
Thanks for reaching out @Sachyyy.
You can close the issue at your convenience.

**Comment by saikiranakula-amzn** [comment]
Now we have a working environment. Let us know if you guys have any more questions. Thank you :)

**Comment by smhmhmd** [comment]
Thanks @Sachyyy  for the discussion + contribution.
